providers/gitlab.go

```diff
@@ -6,199 +6,196 @@ import (
-	return &GitlabProject{
-			Name:        project,
-			AccessLevel: defaultAccessLevel},
-		nil
-
-}
-
-var _ Provider = (*GitLabProvider)(nil)
-
-const (
-	gitlabProviderName = "GitLab"
-	gitlabDefaultScope = "openid email"
-)
-
-// NewGitLabProvider initiates a new GitLabProvider
-func NewGitLabProvider(p *ProviderData) *GitLabProvider {
-	p.ProviderName = gitlabProviderName
-
-	if p.Scope == "" {
-		p.Scope = gitlabDefaultScope
-	}
-
-	return &GitLabProvider{ProviderData: p}
-}
-
-// Redeem exchanges the OAuth2 authentication token for an ID token
-func (p *GitLabProvider) Redeem(ctx context.Context, redirectURL, code string) (s *sessions.SessionState, err error) {
-	clientSecret, err := p.GetClientSecret()
-	if err != nil {
-		return
-	}
-
-	c := oauth2.Config{
-		ClientID:     p.ClientID,
-		ClientSecret: clientSecret,
-		Endpoint: oauth2.Endpoint{
-			TokenURL: p.RedeemURL.String(),
-		},
-		RedirectURL: redirectURL,
-	}
-	token, err := c.Exchange(ctx, code)
-	if err != nil {
-		return nil, fmt.Errorf("token exchange: %v", err)
-	}
-	s, err = p.createSession(ctx, token)
-	if err != nil {
-		return nil, fmt.Errorf("unable to update session: %v", err)
-	}
-	return
+	return &gitlabProject{
+		Name:        project,
+		AccessLevel: defaultAccessLevel,
+	}, nil
 }
 
-// SetProjectScope ensure read_api is added to scope when filtering on projects
-func (p *GitLabProvider) SetProjectScope() {
-	if len(p.Projects) > 0 {
-		for _, val := range strings.Split(p.Scope, " ") {
-			if val == "read_api" {
-				return
-			}
-
+// setProjectScope ensures read_api is added to scope when filtering on projects
+func (p *GitLabProvider) setProjectScope() {
+	for _, val := range strings.Split(p.Scope, " ") {
+		if val == "read_api" {
+			return
 		}
-		p.Scope += " read_api"
 	}
+	p.Scope += " read_api"
 }
 
-// RefreshSessionIfNeeded checks if the session has expired and uses the
-// RefreshToken to fetch a new ID token if required
-func (p *GitLabProvider) RefreshSessionIfNeeded(ctx context.Context, s *sessions.SessionState) (bool, error) {
-	if s == nil || (s.ExpiresOn != nil && s.ExpiresOn.After(time.Now())) || s.RefreshToken == "" {
-		return false, nil
+// EnrichSession enriches the session with the response from the userinfo API
+// endpoint & projects API endpoint for allowed projects.
+func (p *GitLabProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	// Retrieve user info
+	userinfo, err := p.getUserinfo(ctx, s)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve user info: %v", err)
 	}
 
-	origExpiration := s.ExpiresOn
-
-	err := p.redeemRefreshToken(ctx, s)
-	if err != nil {
-		return false, fmt.Errorf("unable to redeem refresh token: %v", err)
+	// Check if email is verified
+	if !p.AllowUnverifiedEmail && !userinfo.EmailVerified {
+		return fmt.Errorf("user email is not verified")
 	}
 
-	logger.Printf("refreshed id token %s (expired on %s)\n", s, origExpiration)
-	return true, nil
-}
+	s.User = userinfo.Username
```

Comment:
What does a GitLab IDToken look like?

Now that this is OIDC based, this method to extract claims into sessions will be called during the OAuth flow before `EnrichSession` ProfileURL calls to extend:

https://github.com/oauth2-proxy/oauth2-proxy/blob/a296936a0f885900cb2db02d2550a76f349f5ef3/providers/provider_data.go#L143

And the OIDCProvider `CreateSessionFromToken` this gets access to for Bearer auth (instead of default implementation) also will use that function to build a richer `Session`

Reply:
Here is the content of an ID token from Gitlab API : 

claim | type | description
-- | -- | --
sub | string | An opaque token that uniquely identifies the user
sub_legacy | string | An opaque token that uniquely identifies the userDeprecation notice: this token isn’t stable because it’s tied to the Rails secret key base, and is provided only for migration to the new stable sub value available from GitLab 11.1
auth_time | integer | The timestamp for the user’s last authentication
name | string | The user’s full name
nickname | string | The user’s GitLab username
email | string | The user’s email addressThis is the user’s primary email address if the application has access to the email claim and the user’s public email address otherwise
email_verified | boolean | Whether the user’s email address was verified
website | string | URL for the user’s website
profile | string | URL for the user’s GitLab profile
picture | string | URL for the user’s GitLab avatar
groups | array | Names of the groups the user is a member of

[Source](https://docs.gitlab.com/ee/integration/openid_connect_provider.html)

Since projects ownership data is not passed we still need to check if the requesting user is member of an authorized project from the configured list.